### PR TITLE
Added recursive binary search

### PR DESCRIPTION
**Pull request by Ursanon**
Added recursive binary search.

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] (NA) I have added comments to hard-to-understand areas of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (NA) I have made corresponding changes to the comments
- [x] I have made corresponding changes to the README.md

### CODE REVIEW

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Algorithms/Search/BinarySearcher.cs  1
         

Clones added
============
- Algorithms.Tests/Search/BinarySearcherTests.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772&bid=20455366)

File: Algorithms/Search/BinarySearcher.cs [hunk 1]
```diff
@@ -43,5 +44,53 @@ public int FindIndex(T[] sortedData, T item)
...
+        /// time complexity: O(log(n)),
+        /// space complexity: O(1),
+        /// where n - collection size.
+        /// </summary>
+        /// <param name="collection">Sorted collection to search in.</param>
+        /// <param name="item">Item to search for.</param>
+        /// <returns>Index of item that equals to item searched for or -1 if none found.</returns>
+        public int FindIndexRecursive(IList<T>? collection, T item)
+        {
+            collection = collection ?? throw new ArgumentNullException(nameof(collection));
```
Comment (siriak) [inline Algorithms/Search/BinarySearcher.cs:59]: Codacy found an issue: [Use the '??=' operator here.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772)

Comment (Ursanon) [reply Algorithms/Search/BinarySearcher.cs:59]: A throw expression is not allowed in this context.

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms/Search/BinarySearcher.cs  1
         

Clones added
============
- Algorithms.Tests/Search/BinarySearcherTests.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772&bid=20455366)

File: Algorithms/Search/BinarySearcher.cs [hunk 2]
```diff
@@ -43,5 +44,56 @@ public int FindIndex(T[] sortedData, T item)
...
+            }
+
+            var middleIndex = leftIndex + (rightIndex - leftIndex) / 2;
+            var result = item.CompareTo(collection[middleIndex]);
+
+            return result switch
+            {
+                var c when c == 0 => middleIndex,
+                var c when c > 0 => FindIndexRecursive(collection, item, middleIndex + 1, rightIndex),
+                _ => FindIndexRecursive(collection, item, leftIndex, middleIndex - 1),
```
Comment (siriak) [inline Algorithms/Search/BinarySearcher.cs:95]: var c when c < 0 ?

**Review by siriak — COMMENTED** [review]
Please move it to a separate RecursiveBinarySearcher class with separate tests

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- Algorithms/Search/RecursiveBinarySearcher.cs  6
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  1
- Algorithms/Search/BinarySearcher.cs  1
         

Clones added
============
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  3
- Algorithms.Tests/Search/BinarySearcherTests.cs  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772&bid=20455366)

File: Algorithms/Search/RecursiveBinarySearcher.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Algorithms.Searches
+{
+    /// <summary>
+    /// TODO.
```
Comment (siriak) [inline Algorithms/Search/RecursiveBinarySearcher.cs:7]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772)

File: Algorithms/Search/RecursiveBinarySearcher.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    /// </summary>
+    /// <typeparam name="T">TODO. 2.</typeparam>
```
Comment (siriak) [inline Algorithms/Search/RecursiveBinarySearcher.cs:9]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772)

File: Algorithms/Search/RecursiveBinarySearcher.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    public class RecursiveBinarySearcher<T> where T : IComparable<T>
+    {
+        public static readonly int NotFoundIndex = -1;
```
Comment (siriak) [inline Algorithms/Search/RecursiveBinarySearcher.cs:12]: Codacy found an issue: [A static field in a generic type is not shared among instances of different close constructed types.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772)

Comment (siriak) [reply Algorithms/Search/RecursiveBinarySearcher.cs:12]: Maybe making it a constant helps?

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  1
- Algorithms/Search/BinarySearcher.cs  1
- Algorithms/Search/RecursiveBinarySearcher.cs  6
         

Clones added
============
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  3
- Algorithms.Tests/Search/BinarySearcherTests.cs  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772&bid=20455366)

File: Algorithms/Search/RecursiveBinarySearcher.cs (hunk 3 above)
Comment (siriak) [inline Algorithms/Search/RecursiveBinarySearcher.cs:7]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772)

File: Algorithms/Search/RecursiveBinarySearcher.cs (hunk 4 above)
Comment (siriak) [inline Algorithms/Search/RecursiveBinarySearcher.cs:9]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772)

File: Algorithms/Search/RecursiveBinarySearcher.cs (hunk 5 above)
Comment (siriak) [inline Algorithms/Search/RecursiveBinarySearcher.cs:12]: Codacy found an issue: [A static field in a generic type is not shared among instances of different close constructed types.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Algorithms/Search/RecursiveBinarySearcher.cs  6
- Algorithms/Search/BinarySearcher.cs  1
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  1
         

Clones added
============
- Algorithms.Tests/Search/BinarySearcherTests.cs  7
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772&bid=20455366)

File: Algorithms/Search/RecursiveBinarySearcher.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace Algorithms.Searches
+{
+    /// <summary>
+    /// RecursiveBinarySearcher
+    /// </summary>
+    /// <typeparam name="T">Type of searcher target</typeparam>
+    public class RecursiveBinarySearcher<T> where T : IComparable<T>
+    {
+        public static readonly int NotFoundIndex = -1;
```
Comment (siriak) [inline Algorithms/Search/RecursiveBinarySearcher.cs:12]: Codacy found an issue: [A static field in a generic type is not shared among instances of different close constructed types.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Algorithms/Search/BinarySearcher.cs  1
- Algorithms/Search/RecursiveBinarySearcher.cs  6
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  1
         

Clones added
============
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  3
- Algorithms.Tests/Search/BinarySearcherTests.cs  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772&bid=20455366)

File: Algorithms/Search/RecursiveBinarySearcher.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace Algorithms.Searches
+{
+    /// <summary>
+    /// RecursiveBinarySearcher
+    /// </summary>
+    /// <typeparam name="T">Type of searcher target</typeparam>
+    public class RecursiveBinarySearcher<T> where T : IComparable<T>
+    {
+        public const int NotFoundIndex = -1;
```
Comment (siriak) [inline Algorithms/Search/RecursiveBinarySearcher.cs:12]: Codacy found an issue: [Change this constant to a 'static' read-only property.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  1
- Algorithms/Search/RecursiveBinarySearcher.cs  6
- Algorithms/Search/BinarySearcher.cs  1
         

Clones added
============
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  3
- Algorithms/Search/RecursiveBinarySearcher.cs  1
- Algorithms/Search/BinarySearcher.cs  1
- Algorithms.Tests/Search/BinarySearcherTests.cs  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772&bid=20455366)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms/Search/RecursiveBinarySearcher.cs  6
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  1
         

Clones added
============
- Algorithms.Tests/Search/BinarySearcherTests.cs  2
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772&bid=20455366)

**Review by siriak — APPROVED** [review]
Could you make a little change anywhere so that Codacy reruns its test and I think it should pass?

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Algorithms/Search/RecursiveBinarySearcher.cs  6
- Algorithms.Tests/Search/RecursiveBinarySearcherTests.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6286772&bid=20455366)

**Review by siriak — APPROVED** [review]
Thanks!

**Pull request merged**